python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +210,221 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+                f"Got {list(self.datasets.keys())}"
+            )
+        if not all(
+            key in (TRAIN_DATASET_KEY, EVALUATION_DATASET_KEY) for key in self.datasets
+        ):
+            raise KeyError(
+                f"Only '{TRAIN_DATASET_KEY}' and '{EVALUATION_DATASET_KEY}' "
+                "keys can be preset in `datasets`. "
+                f"Got {list(self.datasets.keys())}"
+            )
+        gpus_per_worker = self.scaling_config.get("num_gpus_per_worker", 0)
+        if gpus_per_worker > 1:
+            raise ValueError(
+                f"You have assigned {gpus_per_worker} GPUs per worker. "
+                "This is not supported by HuggingFace, which expects "
+                "one GPU per worker in DDP mode and will fail "
+                "if more are assigned."
+            )
+        if gpus_per_worker != int(gpus_per_worker):
+            raise ValueError(
+                f"You have assigned {gpus_per_worker} GPUs per worker, "
+                "but fractional GPUs are not supported by HuggingFace."
+            )
+
+        super()._validate_attributes()
+
+    def _convert_directory_checkpoint_to_sync(
+        self, checkpoint: Checkpoint
+    ) -> Checkpoint:
+        """Replace the directory checkpoint with a node ip & path dict checkpoint
+        used to sync the directory. If we use a directory checkpoint directly,
+        it will get deepcopied & serialized unnecessarily."""
+        with checkpoint.as_directory() as checkpoint_path:
+            # Load checkpoint from path.
+            checkpoint_path = Path(checkpoint_path).expanduser().absolute()
+            if not checkpoint_path.exists():
+                raise ValueError(f"Checkpoint path {checkpoint_path} does not exist.")
+            with open(checkpoint_path.joinpath(TUNE_CHECKPOINT_ID), "r") as f:
+                tune_checkpoint_id = int(f.read())
+
+            return Checkpoint.from_dict(
+                {
+                    NODE_IP_KEY: get_node_ip_address(),
+                    CHECKPOINT_PATH_ON_NODE_KEY: str(checkpoint_path),
+                    TUNE_CHECKPOINT_ID: tune_checkpoint_id,
+                }
+            )
+
+    def setup(self) -> None:
+        if (
+            self.resume_from_checkpoint
+            and self.resume_from_checkpoint.get_internal_representation()
+            == "local_path"
+        ):
+            self.resume_from_checkpoint = self._convert_directory_checkpoint_to_sync(
+                self.resume_from_checkpoint
+            )
+
+    def as_trainable(self) -> Type[Trainable]:
+        original_param_dict = self._param_dict.copy()
+        resume_from_checkpoint: Optional[Checkpoint] = self._param_dict.get(
+            "resume_from_checkpoint", None
         )
+        if resume_from_checkpoint:
+            self._param_dict[
+                "resume_from_checkpoint"
+            ] = self._convert_directory_checkpoint_to_sync(resume_from_checkpoint)
```

Comment:
Ah I see we need this in both `as_trainable` and `setup` since we have to support both user-specified resuming and Tune resuming. 

If we change the logic here a little bit: https://github.com/ray-project/ray/blob/master/python/ray/ml/trainer.py#L347-L352

to be like this instead 

```python
if checkpoint_dir:
    config["resume_from_checkpoint"] = Checkpoint.from_directory(checkpoint_dir)
trainer = trainer_cls(**config)
```

then `_convert_directory_checkpoint_to_sync` only needs to be called in init, right?

Reply:
We still need to convert it to the sync checkpoint format, and that has to happen before it's sent to the Actor in case of user specified resume. We basically have two cases here, as you noticed:
* In case of user specified resume, we convert the dir checkpoint to sync checkpoint on the driver and send over the sync checkpoint
* In case of tune resume, we convert the dir checkpoint to sync checkpoint in the Actor

We need to end up with a sync checkpoint in both cases, as that's what's expected by the checkpoint manager/syncing logic.

Reply:
Correct— my suggestion would still resolve both cases right?

Reply:
That should work, yeah. Not sure if this change should be made here though, as it would impact all trainers (possibly in unforeseen ways). Let's do that in a followup PR